File: 30 Days of Code/Day 24 More Linked Lists/remove_duplicate.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def Display(self):
        current = self.head

        while current != None:
            print(str(current.data) + ' ', end = '')
            current = current.next

    def Insert(self, data):
        if self.head == None:
            self.head = Node(data)
        else:
            current = self.head

            while current.next != None:
                current = current.next

            current.next = Node(data)

    def RemoveDuplicate(self):
        current = self.head.next
        previous = self.head

        while current != None:
            if previous.data != current.data:
                previous.next = current
                previous = current
            current = current.next
        previous.next = None


n = int(input())
llist = LinkedList()
for i in range(n):
    llist.Insert(int(input()))

llist.RemoveDuplicate()
llist.Display()

File: 30 Days of Code/Day 24 More Linked Lists/remove_duplicate.cpp
#include <iostream>
#include "remove_duplicate.hpp"
using namespace std;

template <class T>
Node<T>::Node(T data):
	data_(data),
	next_(nullptr)
{}

// Delete next call its next deleted in order.
template <class T>
Node<T>::~Node()
{
	delete next_;
}

template <class T>
T Node<T>::GetData()
{
	return data_;
}

template <class T>
Node<T> *Node<T>::GetNext()
{
	return next_;
}

template <class T>
void Node<T>::SetData(T data)
{
	data_ = data;
}

template <class T>
void Node<T>::SetNext(Node<T> *next)
{
	next_ = next;
}

template <class T>
LinkedList<T>::LinkedList():
	head_(nullptr)
{}

template <class T>
LinkedList<T>::~LinkedList()
{
	delete head_;
}

template <class T>
void LinkedList<T>::Display()
{
	Node<T> *current = head_;

	while (current != nullptr) {
		cout << current->GetData() << " ";
		current = current->GetNext();
	}
}

template <class T>
void LinkedList<T>::Insert(T data)
{
	if (head_ == nullptr) {
		head_ = new Node<T>(data);
	} else {
		Node<T> *current = head_;

		while (current->GetNext() != nullptr) {
			current = current->GetNext();
		}

		current->SetNext(new Node<T>(data));
	}
}

template <class T>
void LinkedList<T>::RemoveDuplicate()
{
	Node<T> *current = head_->GetNext();
	Node<T> *previous = head_;

	while (current != nullptr) {
		if (previous->GetData() != current->GetData()) {
			previous->SetNext(current);
			previous = current;
		}
		current = current->GetNext();
	}
	previous->SetNext(nullptr);
}

int main()
{
	int n = 0;
	cin >> n;
	LinkedList<int> *llist = new LinkedList<int>();

	for (int i = 0; i < n; i++) {
		int num = 0;
		cin >> num;
		llist->Insert(num);
	}

	llist->RemoveDuplicate();
	llist->Display();
	delete llist;
}